frontend/src/metabase/containers/AddToDashSelectDashModal.tsx

```diff
@@ -0,0 +1,104 @@
+/* eslint-disable import/order */
+/* eslint-disable react/prop-types */
+import React, { Component, ComponentPropsWithoutRef } from "react";
+import { connect } from "react-redux";
+import { t } from "ttag";
+
+import Icon from "metabase/components/Icon";
+import Link from "metabase/core/components/Link";
+import ModalContent from "metabase/components/ModalContent";
+import DashboardPicker from "metabase/containers/DashboardPicker";
+import type { State } from "metabase-types/store";
+import { CreateDashboardFormOwnProps } from "metabase/dashboard/containers/CreateDashboardForm";
+
+import * as Urls from "metabase/lib/urls";
+
+import CreateDashboardModal from "metabase/dashboard/containers/CreateDashboardModal";
+
+import { LinkContent } from "./AddToDashSelectDashModal.styled";
+import { Card, Dashboard } from "metabase-types/api";
+
+function mapStateToProps(state: State) {
+  return {
+    dashboards: state.entities.dashboards,
+  };
+}
+
+interface AddToDashSelectDashModalState {
+  shouldCreateDashboard: boolean;
+}
+
+interface AddToDashSelectDashModalProps {
+  card: Card;
+  onChangeLocation: (location: string) => void;
+  onClose: () => void;
+  dashboards: Record<Dashboard["id"], Dashboard>;
+}
+
+type DashboardPickerProps = ComponentPropsWithoutRef<typeof DashboardPicker>;
+
+class AddToDashSelectDashModal extends Component<
+  AddToDashSelectDashModalProps,
+  AddToDashSelectDashModalState
+> {
+  state = {
+    shouldCreateDashboard: false,
+  };
+
+  // ugly hack with the type, could be better
+  navigateToDashboard: Required<CreateDashboardFormOwnProps>["onCreate"] =
+    dashboard => {
+      const { card, onChangeLocation } = this.props;
+
+      onChangeLocation(
+        Urls.dashboard(dashboard, {
+          editMode: true,
+          addCardWithId: card.id,
+        }),
+      );
+    };
+
+  onDashboardSelected: DashboardPickerProps["onChange"] = dashboardId => {
+    const dashboard = this.props.dashboards[dashboardId as Dashboard["id"]];
+    this.navigateToDashboard(dashboard);
+  };
+
+  render() {
+    if (this.state.shouldCreateDashboard) {
+      return (
+        <CreateDashboardModal
+          // @ts-expect-error collection_id can be null?
+          collectionId={this.props.card.collection_id}
+          onCreate={this.navigateToDashboard}
+          onClose={() => this.setState({ shouldCreateDashboard: false })}
+        />
+      );
+    }
```

Comment:
I think we don't need a `@ts-expect-error` here. Canonical root collection's ID is `null`, so the types should respect that

Reply:
the beauty of `ts-expect-error` is that it will fail as soon as there is no ts error, I even have a PR about it https://github.com/metabase/metabase/pull/29796

probably I can create RFC explaining this, should I?

Reply:
Yeah, I think if there's a way to avoid using it, we should go for it — collection ID can be either a number or a null (also could be "root" when we need an ID-collection map), so there should be a way to properly type this

Re: RFC — ofc!